### PR TITLE
Add show_create_query_identifier_quoting_rule setting

### PR DESCRIPTION
**Pull request by tuanpach**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add `show_create_query_identifier_quoting_rule` to define identifier quoting behavior of the show create query result. Possible values:
- `user_display`: When the identifiers is a keyword.
- `when_necessary`: When the identifiers is one of `{"distinct", "all", "table"}`, or it can cause ambiguity: column names, dictionary attribute names.
- `always`: Always quote identifiers

Rename `output_format_identifier_quoting_style` to `show_create_query_identifier_quoting_style`, and remove `IdentifierQuotingStyle::None`

Remove `output_format_always_quote_identifiers` setting.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit 07aee9d8afa0577784ed097630c8615271e6336f with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/69448/07aee9d8afa0577784ed097630c8615271e6336f/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/db42cf12fd0e7a54b9a63118b21c300eddb7970e/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/Parsers/ASTAssignment.h [hunk 1]
```diff
@@ -30,7 +30,7 @@ class ASTAssignment : public IAST
     {
 
         settings.ostr << (settings.hilite ? hilite_identifier : "");
-        settings.writeIdentifier(column_name);
+        settings.writeIdentifier(column_name, false);
```
Comment (tavplubix) [inline src/Parsers/ASTAssignment.h:33]:
```suggestion
        settings.writeIdentifier(column_name, /*ambiguous*/ false);
```

File: src/Parsers/IdentifierQuotingStyle.h [hunk 2]
```diff
@@ -14,4 +14,10 @@ enum class IdentifierQuotingStyle : uint8_t
     BackticksMySQL,  /// `mysql` style, most same as Backticks, but it uses '``' to escape '`'
 };
 
+enum class IdentifierQuotingRule : uint8_t
+{
+    UserDisplay,    /// When the identifiers is one of the certain keywords defined in `writeProbablyQuotedStringImpl`
+    WhenNecessary,  /// When the identifiers is one of the certain keywords defined in `writeProbablyQuotedStringImpl`, and ambiguous identifiers passed to `writeIdentifier`
+    Always,         /// Always quote identifiers
```
Comment (tavplubix) [inline src/Parsers/IdentifierQuotingStyle.h:21]: We should add these descriptions to the [documentation](https://clickhouse.com/docs/en/operations/settings/settings) as well (see settings.md)

File: src/Parsers/IAST.cpp [hunk 3]
```diff
@@ -217,38 +218,44 @@ String IAST::getColumnNameWithoutAlias() const
 }
 
 
-void IAST::FormatSettings::writeIdentifier(const String & name) const
+void IAST::FormatSettings::writeIdentifier(const String & name, bool ambiguous) const
```
Comment (tavplubix) [inline src/Parsers/IAST.cpp:221]: Consider searching for this identifier in the [list of keywords](https://github.com/ClickHouse/ClickHouse/blob/129d1e387e1b436ac572ca5db1a50dd9ed79ddae/src/Parsers/CommonParsers.h#L588) to check if it's ambiguous or not

Comment (tuanpach) [reply src/Parsers/IAST.cpp:224]: I already updated the PR to check if the identifier is one of the keywords.
Some keywords (I believe) are commonly used as identifiers `date`, `day`, `min`, `max`, `MONTH`, etc. So they will be quoted by default.

File: docs/en/operations/settings/settings.md [hunk 4]
```diff
@@ -5683,6 +5683,15 @@ Enable `IF NOT EXISTS` for `CREATE` statement by default. If either this setting
 
 Default value: `false`.
 
+## output_format_identifier_quoting_rule
+
+Define identifier quoting behavior:
+- `when_necessary`: When the identifiers is one of `{"distinct", "all", "table"}` (defined in `DB::writeProbablyQuotedStringImpl`), and ambiguous identifiers: column names, dictionary attribute names (passed to `DB::FormatSettings::writeIdentifier` with `ambiguous=true`)
```
[the remark this answers is not in the document]
Comment (tuanpach) [reply docs/en/operations/settings/settings.md]: We quote 3 keywords and identifiers that cause ambiguity: column names + dictionary attribute names.

**Review by tavplubix — APPROVED** [review]

**Pull request merged**